File: Services/FeishuBridge/lib/inbound-media.js
const crypto = require('node:crypto');
const fs = require('node:fs');
const path = require('node:path');
const { chmodPrivate } = require('./platform-runtime');

const SUPPORTED_INBOUND_MESSAGE_TYPES = new Set(['text', 'image', 'file', 'audio', 'media', 'post']);

function parseContent(content) {
  if (content && typeof content === 'object') return content;
  try {
    return JSON.parse(String(content || '{}'));
  } catch {
    return {};
  }
}

function safeFileName(value, fallback) {
  const base = path.basename(String(value || '')).normalize('NFKC')
    .replace(/[\u0000-\u001f\u007f]/g, '')
    .replace(/[\\/:*?"<>|]/g, '_')
    .replace(/^\.+/, '')
    .trim();
  return (base || fallback).slice(0, 160);
}

function postDetails(content) {
  const textParts = [];
  const resources = [];
  const seen = new Set();
  function visit(value) {
    if (Array.isArray(value)) {
      value.forEach(visit);
      return;
    }
    if (!value || typeof value !== 'object') return;
    if (value.tag === 'text' && typeof value.text === 'string') textParts.push(value.text);
    if (typeof value.title === 'string') textParts.push(value.title);
    for (const [key, resourceType] of [['image_key', 'image'], ['file_key', 'file']]) {
      if (typeof value[key] !== 'string' || seen.has(value[key])) continue;
      seen.add(value[key]);
      resources.push({
        fileKey: value[key],
        resourceType,
        displayName: safeFileName(value.file_name, `${resourceType}-${resources.length + 1}`),
      });
    }
    Object.values(value).forEach(visit);
  }
  visit(content);
  return { text: textParts.join('\n').trim(), resources };
}

function inboundMessageDetails(message) {
  const messageType = String(message?.message_type || '');
  const content = parseContent(message?.content);
  if (messageType === 'text') return { messageType, text: String(content.text || ''), resources: [] };
  if (messageType === 'post') return { messageType, ...postDetails(content) };
  if (messageType === 'image' && content.image_key) {
    return {
      messageType,
      text: '',
      resources: [{ fileKey: content.image_key, resourceType: 'image', displayName: 'image' }],
    };
  }
  if (['file', 'audio', 'media'].includes(messageType) && content.file_key) {
    return {
      messageType,
      text: String(content.text || ''),
      resources: [{
        fileKey: content.file_key,
        resourceType: 'file',
        displayName: safeFileName(content.file_name, `${messageType}-file`),
      }],
    };
  }
  return { messageType, text: '', resources: [] };
}

function inboundRoot(cwd) {
  return path.join(path.resolve(cwd), 'node_modules', '.cache', 'feishu-bridge', 'inbound-assets');
}

function assertScopedDirectory(cwd, directory) {
  const root = inboundRoot(cwd);
  const relative = path.relative(root, path.resolve(directory));
  if (!relative || relative.startsWith('..') || path.isAbsolute(relative)) {
    throw new Error('unsafe inbound asset cleanup path');
  }
  return root;
}

function cleanupInboundAssets(cwd, directory) {
  if (!directory) return;
  assertScopedDirectory(cwd, directory);
  fs.rmSync(directory, { recursive: true, force: true });
}

function regularFiles(directory) {
  if (!fs.existsSync(directory)) return [];
  return fs.readdirSync(directory).filter((name) => {
    const stat = fs.lstatSync(path.join(directory, name));
    return stat.isFile() && !stat.isSymbolicLink();
  });
}

async function stageInboundMessage({
  lark,
  message,
  cwd,
  maxBytes = 25 * 1024 * 1024,
  maxResources = 10,
}) {
  const details = inboundMessageDetails(message);
  if (!SUPPORTED_INBOUND_MESSAGE_TYPES.has(details.messageType)) {
    throw new Error(`unsupported inbound message type: ${details.messageType || '-'}`);
  }
  if (details.resources.length > maxResources) throw new Error('too many inbound resources');
  if (!details.resources.length) return { ...details, assets: [], cleanupDir: '' };

  const root = inboundRoot(cwd);
  fs.mkdirSync(root, { recursive: true, mode: 0o700 });
  const rootStat = fs.lstatSync(root);
  if (rootStat.isSymbolicLink() || !rootStat.isDirectory()) throw new Error('unsafe inbound asset root');
  const realRoot = fs.realpathSync(root);
  const realCwd = fs.realpathSync(path.resolve(cwd));
  const rootRelative = path.relative(realCwd, realRoot);
  if (!rootRelative || rootRelative.startsWith('..') || path.isAbsolute(rootRelative)) {
    throw new Error('inbound asset root escapes the bridge project');
  }
  chmodPrivate(root, 0o700);
  const digest = crypto.createHash('sha256').update(String(message.message_id || '')).digest('hex').slice(0, 24);
  const cleanupDir = path.join(root, digest);
  cleanupInboundAssets(cwd, cleanupDir);
  fs.mkdirSync(cleanupDir, { mode: 0o700 });

  const assets = [];
  let totalBytes = 0;
  try {
    for (const [index, resource] of details.resources.entries()) {
      const outputName = `${String(index + 1).padStart(2, '0')}-${safeFileName(resource.displayName, 'resource')}`;
      const outputPath = path.join(cleanupDir, outputName);
      const before = new Set(regularFiles(cleanupDir));
      const relativeOutput = path.relative(path.resolve(cwd), outputPath).split(path.sep).join('/');
      await lark.larkImDownloadMessageResource({
        messageId: message.message_id,
        fileKey: resource.fileKey,
        resourceType: resource.resourceType,
        output: relativeOutput,
        timeoutMs: 120000,
      });
      const candidates = regularFiles(cleanupDir).filter((name) => !before.has(name));
      const downloaded = fs.existsSync(outputPath) ? outputName : candidates[0];
      if (!downloaded) throw new Error('resource download did not create a regular file');
      const localPath = path.join(cleanupDir, downloaded);
      const stat = fs.lstatSync(localPath);
      if (stat.isSymbolicLink() || !stat.isFile()) throw new Error('downloaded resource is not a regular file');
      totalBytes += stat.size;
      if (totalBytes > maxBytes) throw new Error('inbound resources exceed size limit');
      chmodPrivate(localPath, 0o600);
      assets.push({
        messageType: details.messageType,
        resourceType: resource.resourceType,
        displayName: safeFileName(resource.displayName, downloaded),
        localPath,
        sizeBytes: stat.size,
      });
    }
    return { ...details, assets, cleanupDir };
  } catch (error) {
    cleanupInboundAssets(cwd, cleanupDir);
    throw error;
  }
}

function inboundAssetsPrompt(prompt, details) {
  if (!details?.assets?.length) return String(prompt || '');
  const files = details.assets.map((asset, index) => (
    `${index + 1}. ${asset.displayName} (${asset.messageType}/${asset.resourceType}, ${asset.sizeBytes} bytes): ${asset.localPath}`
  ));
  return [
    String(prompt || '').trim() || '请分析我发送的附件并给出有用的回复。',
    '',
    '飞书桥已将本轮附件暂存为以下本机只读输入。请按用户意图读取这些文件；不要移动、改写或删除它们：',
    ...files,
  ].join('\n');
}

module.exports = {
  SUPPORTED_INBOUND_MESSAGE_TYPES,
  cleanupInboundAssets,
  inboundAssetsPrompt,
  inboundMessageDetails,
  safeFileName,
  stageInboundMessage,
};

File: Core/internal/feishu/inbound_media.go
package feishu

import (
	"context"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"
	"unicode"
)

const (
	defaultInboundMaxBytes     = 25 * 1024 * 1024
	defaultInboundMaxResources = 10
)

type InboundResource struct {
	FileKey, ResourceType, DisplayName string
}

type InboundAsset struct {
	MessageType, ResourceType, DisplayName, LocalPath string
	SizeBytes                                         int64
}

type StagedInboundMessage struct {
	MessageType string
	Text        string
	Assets      []InboundAsset
	CleanupDir  string
}

func inboundMessageDetails(message InboundMessage) (string, []InboundResource, error) {
	var content any
	rawMessage, _ := message.Raw["message"].(map[string]any)
	if err := json.Unmarshal([]byte(stringValue(rawMessage["content"])), &content); err != nil {
		content = map[string]any{}
	}
	switch message.MessageType {
	case "text":
		return message.Text, nil, nil
	case "image":
		value, _ := content.(map[string]any)
		key := stringValue(value["image_key"])
		if key == "" {
			return "", nil, errors.New("inbound_image_key_missing")
		}
		return "", []InboundResource{{FileKey: key, ResourceType: "image", DisplayName: "image"}}, nil
	case "file", "audio", "media":
		value, _ := content.(map[string]any)
		key := stringValue(value["file_key"])
		if key == "" {
			return "", nil, errors.New("inbound_file_key_missing")
		}
		return stringValue(value["text"]), []InboundResource{{FileKey: key, ResourceType: "file", DisplayName: safeInboundFileName(stringValue(value["file_name"]), message.MessageType+"-file")}}, nil
	case "post":
		text := []string{}
		resources := []InboundResource{}
		seen := map[string]bool{}
		var visit func(any)
		visit = func(value any) {
			switch item := value.(type) {
			case []any:
				for _, child := range item {
					visit(child)
				}
			case map[string]any:
				if item["tag"] == "text" && stringValue(item["text"]) != "" {
					text = append(text, stringValue(item["text"]))
				}
				if title := stringValue(item["title"]); title != "" {
					text = append(text, title)
				}
				for key, kind := range map[string]string{"image_key": "image", "file_key": "file"} {
					fileKey := stringValue(item[key])
					if fileKey != "" && !seen[fileKey] {
						seen[fileKey] = true
						resources = append(resources, InboundResource{FileKey: fileKey, ResourceType: kind, DisplayName: safeInboundFileName(stringValue(item["file_name"]), fmt.Sprintf("%s-%d", kind, len(resources)+1))})
					}
				}
				for _, child := range item {
					visit(child)
				}
			}
		}
		visit(content)
		return strings.TrimSpace(strings.Join(text, "\n")), resources, nil
	default:
		return "", nil, fmt.Errorf("unsupported inbound message type: %s", message.MessageType)
	}
}

func StageInboundMessage(ctx context.Context, runner CapabilityExecutor, message InboundMessage, maxBytes int64) (StagedInboundMessage, error) {
	text, resources, err := inboundMessageDetails(message)
	if err != nil {
		return StagedInboundMessage{}, err
	}
	if len(resources) > defaultInboundMaxResources {
		return StagedInboundMessage{}, errors.New("too_many_inbound_resources")
	}
	if len(resources) == 0 {
		return StagedInboundMessage{MessageType: message.MessageType, Text: text, Assets: []InboundAsset{}}, nil
	}
	if maxBytes <= 0 {
		maxBytes = defaultInboundMaxBytes
	}
	root := filepath.Join(runner.DataRoot, "private-cache", "inbound-assets")
	if err := ensurePrivateDirectory(root); err != nil {
		return StagedInboundMessage{}, err
	}
	digest := sha256.Sum256([]byte(message.MessageID))
	directory := filepath.Join(root, hex.EncodeToString(digest[:12]))
	if err := CleanupInboundAssets(runner.DataRoot, directory); err != nil {
		return StagedInboundMessage{}, err
	}
	if err := os.Mkdir(directory, 0o700); err != nil {
		return StagedInboundMessage{}, err
	}
	result := StagedInboundMessage{MessageType: message.MessageType, Text: text, Assets: []InboundAsset{}, CleanupDir: directory}
	failed := true
	defer func() {
		if failed {
			_ = CleanupInboundAssets(runner.DataRoot, directory)
		}
	}()
	var total int64
	for index, resource := range resources {
		name := fmt.Sprintf("%02d-%s", index+1, safeInboundFileName(resource.DisplayName, "resource"))
		output := filepath.Join(directory, name)
		if err := runner.DownloadMessageResource(ctx, message.MessageID, resource.FileKey, resource.ResourceType, output, 2*time.Minute); err != nil {
			return StagedInboundMessage{}, err
		}
		info, err := os.Lstat(output)
		if err != nil || !info.Mode().IsRegular() || info.Mode()&os.ModeSymlink != 0 {
			return StagedInboundMessage{}, errors.New("downloaded_resource_is_not_a_regular_file")
		}
		total += info.Size()
		if total > maxBytes {
			return StagedInboundMessage{}, errors.New("inbound_resources_exceed_size_limit")
		}
		if err := os.Chmod(output, 0o600); err != nil {
			return StagedInboundMessage{}, err
		}
		result.Assets = append(result.Assets, InboundAsset{MessageType: message.MessageType, ResourceType: resource.ResourceType, DisplayName: resource.DisplayName, LocalPath: output, SizeBytes: info.Size()})
	}
	failed = false
	return result, nil
}

func InboundAssetsPrompt(prompt string, staged StagedInboundMessage) string {
	if len(staged.Assets) == 0 {
		return prompt
	}
	lines := []string{strings.TrimSpace(prompt), "", "飞书服务已将本轮附件暂存为以下本机只读输入。请按用户意图读取这些文件；不要移动、改写或删除它们："}
	if lines[0] == "" {
		lines[0] = "请分析我发送的附件并给出有用的回复。"
	}
	for index, asset := range staged.Assets {
		lines = append(lines, fmt.Sprintf("%d. %s (%s/%s, %d bytes): %s", index+1, asset.DisplayName, asset.MessageType, asset.ResourceType, asset.SizeBytes, asset.LocalPath))
	}
	return strings.Join(lines, "\n")
}

func CleanupInboundAssets(dataRoot, directory string) error {
	if strings.TrimSpace(directory) == "" {
		return nil
	}
	root := filepath.Join(dataRoot, "private-cache", "inbound-assets")
	relative, err := filepath.Rel(root, directory)
	if err != nil || relative == "." || filepath.IsAbs(relative) || strings.HasPrefix(relative, ".."+string(os.PathSeparator)) {
		return errors.New("unsafe_inbound_asset_cleanup_path")
	}
	return os.RemoveAll(directory)
}

func safeInboundFileName(value, fallback string) string {
	name := filepath.Base(strings.TrimSpace(value))
	name = strings.Map(func(r rune) rune {
		if unicode.IsControl(r) || strings.ContainsRune(`\\/:*?"<>|`, r) {
			return '_'
		}
		return r
	}, name)
	name = strings.TrimLeft(name, ".")
	name = strings.TrimSpace(name)
	if name == "" {
		name = fallback
	}
	runes := []rune(name)
	if len(runes) > 160 {
		name = string(runes[:160])
	}
	return name
}
